keps/sig-cluster-lifecycle/0035-20190128-addons-via-operators.md

```diff
@@ -0,0 +1,233 @@
+  themselves can remain bounded in scope and complexity.  Patching will fail if
+  the underlying addon changes dramatically, so we'll likely have a "patch
+  incompatible with new version" error - and generally addons should avoid
+  gratuitously changing their structure.
+* We should develop a convention (labels / owner-refs) so that we are able to
+  discover which CRDs are cluster-addons, and there is no confusion with
+  application operators.
+
+We expect the following functionality to be common to all operators for addons:
+
+* A CRD per addon
+* Common fields in spec that define the version and/or channel
+* Common fields in status that expose the current health & version information
+  of the addon
+* Addons follow a common structure, with an instance of the CRD as root object,
+  an Application CRD instance, consistent labels of all objects
+* Some form of protection or rapid reconciliation to prevent accidental
+  modification of child objects
+* Operators are declaratively driven, and can source manifests via https
+  (including mirrors), or from data stored in the cluster itself
+  (e.g. configmaps or cluster-bundle CRD, useful for airgapped)
+* Operators are able to expose different update behaviours: automatic immediate
+  updates; notification of update-available in status; purely manual updates
+* Operators are able to observe other CRDs to perform basic sequencing
+* Addon manifests are able express an operator minimum version requirement, so
+  that an addon with new requirements can require that the operator be updated
+  first
+* Airgapped operation should be possible by combining a registry mirror and
+  storage of the underlying manifest in the cluster itself.
+
+
+An example can make this easier to understand, here is what a CRD instance for
+kube-proxy might look like:
+
+```yaml
+apiVersion: addons.sigs.k8s.io/v1alpha1
+kind: KubeProxy
+metadata:
+  name: default
+  namespace: kube-system
+spec:
+  clusterCidr: 100.64.0.0/10
+  version: 1.14.4
+status:
+  healthy: true
+```
+
+This particular manifest is pinned to `version: 1.14.4`.  We could also
+subscribe to a stream of updates with a field like `channel: stable`.
+
+### Risks and Mitigations
+
+This will involve running a large number of new controllers.  This will require
+more resources; we can mitigate this by combining them into a single binary
+(similar to kube-controller-manager).
+
+Automatically updating addons could result in new SPOFs, we can mitigate this
+through mirroring (including support for air-gapped mirrors).
+
+Providing a good set of addons could result in a monoculture where mistakes
+affect most/all kubernetes clusters (even if we don’t mandate adoption, if we
+succeed we hope for widespread adoption).  We can continue with our strategies
+that we use for core components such as kube-apiserver: primarily we must keep
+the notion of stable vs less-stable releases, to stagger the risk of a bad
```

Comment:
I don't think we have the notion of stable vs less stable releases today. Individual components, like Core DNS, have become the default once they are considered GA.

Reply:
Right - I'm saying that we would expect that addon X might continue to support 1.1 for a while after introducing 1.2 to avoid the monoculture risk.  They might choose to leverage channels for this, for example choosing to define a "stable" channel that stays on 1.1 for a while, while the "beta" channel moves to 1.2 until they feel it has been derisked.

Today this work is all on the kubernetes test & release teams, which feels very centralized.